src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs

```diff
@@ -0,0 +1,172 @@
+                            {
+                                if (typeof(IRGRestartGameAction).IsAssignableFrom(type))
+                                {
+                                    RGDebug.LogInfo($"Using the '{type.FullName}' implementation of IRGRestartGameAction to process Bot Restart Game Actions.  If this is not the type you expected to handle this action, you may accidentally have more than one implementation of the IRGRestartGameAction interface in your runtime.");
+                                    _action = (IRGRestartGameAction)Activator.CreateInstance(type);
+                                    break;
+                                }
+                            }
+                        }
+
+                        if (_action == null)
+                        {
+                            _error = $"Regression Games could not find an IRGRestartGameAction implementation. The system will use the default restart action of SceneManager.LoadScene(sceneBuildIndex: 0, mode: LoadSceneMode.Single)";
+                            RGDebug.LogWarning(_error);
+                        }
+                        else
+                        {
+                            _error = null;
+                        }
+                    }).Start();
+#endif
+                }
+            }
+        }
+
+#if UNITY_EDITOR
+        private void PlayGameInEditor(PlayModeStateChange state)
+        {
+            if (state == PlayModeStateChange.EnteredEditMode)
+            {
+                // un-register our listener
+                EditorApplication.playModeStateChanged -= PlayGameInEditor;
+                // just finished playing ... start it back up
+                EditorApplication.isPlaying = true;
+            }
+        }
+#endif
+
+        public bool ProcessAction(int segmentNumber, Dictionary<long, ObjectStatus> currentTransforms, Dictionary<long, ObjectStatus> currentEntities, out string error)
+        {
+            if (!_isStopped)
+            {
+                _isStopped = true;
+                //TODO (REG-2170): Write a status to persistent path so that we can resume this sequence/segment on game restart
+                //TODO (REG-2170): (in code somewhere else ...) If this is NOT the last segment, block the upload from happening yet as we aren't 'done' recording the replay
+                //TODO (REG-2170): (in code somewhere else ...) Read recovery status from persistent path so that we can resume this sequence/segment on game restart
+#if UNITY_EDITOR
+
+                RGDebug.LogInfo($"Restarting the game in the editor...");
+                _error = null;
+                error = _error;
+                // register our hook so that the game will start right back up
+                EditorApplication.playModeStateChanged += PlayGameInEditor;
+                // stop the game in the editor
+                EditorApplication.isPlaying = false;
+                return true;
```

Comment:
correct.. as of right now.. 

- Editor - 'restart' actually restarts the game
- External Runtime - 
  - Default - 'restart' loads scene 0, which is as close as we can get to restarting the game without ending the process runtime; this restarts the game to the starting scene, but potentially without cleaning up many things as noted in the comments in the code... we could explore making this much harsher... like.. destroy all objects (event don't destroy on load) before loading scene 0 which would make it much closer to the editor behaviour... however, this still wouldn't kill background threads, or unity task running on entities/etc
  - Custom - 'restart' runs whatever commands are in your custom impl

Reply:
I suppose that was fine for now, but I was thinking that even in editor it would run the custom impl, since my restart logic might just be "respawn at the respawn point and remove all enemies". But on second thought, it might be good to allow the custom action in the editor because this means the only way for you to test the custom logic is to keep rebuilding the game, which could take a long time.

Reply:
Apparently we weren't on the same page at all when initially discussing this task.

It seems you wanted an action to define 'reset for next test', but I got 'restart game' out of the initial discussion.

I guess I'll start over .

Reply:
Start over? I'm confused as I feel like all the code here aligns with exactly what I was thinking, including the default of loading the initial scene. From what I see here, this aligns almost exactly with my thinking but I had in my mind one small extension, but if that extension is not small I'd rather have this PR in.